File: src/typescript/src/preprocessing.ts
import type { ImageData } from "./io.js";

export interface Samples {
  data: Float32Array;
  rows: number;
  cols: number;
}

export function imageToSamples(image: ImageData): Samples {
  const { data, width, height } = image;
  const rows = width * height;
  const cols = 3;
  const samples = new Float32Array(rows * cols);

  for (let i = 0; i < rows; i += 1) {
    const srcIdx = i * 3;
    const dstIdx = i * cols;
    samples[dstIdx] = data[srcIdx];
    samples[dstIdx + 1] = data[srcIdx + 1];
    samples[dstIdx + 2] = data[srcIdx + 2];
  }

  return { data: samples, rows, cols };
}

export function samplesToImage(
  samples: Samples,
  width: number,
  height: number,
): ImageData {
  if (samples.cols !== 3) {
    throw new Error("samples must have 3 columns");
  }

  if (samples.rows !== width * height) {
    throw new Error("sample count mismatch");
  }

  const data = new Uint8Array(width * height * 3);
  const { rows, cols } = samples;

  for (let i = 0; i < rows; i += 1) {
    const srcIdx = i * cols;
    const dstIdx = i * 3;
    data[dstIdx] = clampToByte(samples.data[srcIdx]);
    data[dstIdx + 1] = clampToByte(samples.data[srcIdx + 1]);
    data[dstIdx + 2] = clampToByte(samples.data[srcIdx + 2]);
  }

  return { data, width, height };
}

function clampToByte(value: number): number {
  if (value < 0) return 0;
  if (value > 255) return 255;
  return value | 0;
}

File: src/rust/src/preprocessing.rs
use anyhow::{ensure, Result};
use image::RgbImage;
use ndarray::Array2;

pub fn image_to_samples(image: &RgbImage) -> Result<Array2<f32>> {
    let (width, height) = (image.width() as usize, image.height() as usize);
    ensure!(width > 0 && height > 0, "image dimensions must be positive");

    let mut samples = Array2::<f32>::zeros((width * height, 3));
    for (idx, pixel) in image.pixels().enumerate() {
        samples[[idx, 0]] = pixel[0] as f32;
        samples[[idx, 1]] = pixel[1] as f32;
        samples[[idx, 2]] = pixel[2] as f32;
    }
    Ok(samples)
}

pub fn samples_to_image(samples: &Array2<f32>, width: u32, height: u32) -> Result<RgbImage> {
    ensure!(samples.ncols() == 3, "samples must have 3 columns");
    ensure!(samples.nrows() == (width * height) as usize, "row count mismatch");

    let mut image = RgbImage::new(width, height);
    for (idx, pixel) in image.pixels_mut().enumerate() {
        let row = samples.row(idx);
        pixel[0] = row[0].clamp(0.0, 255.0) as u8;
        pixel[1] = row[1].clamp(0.0, 255.0) as u8;
        pixel[2] = row[2].clamp(0.0, 255.0) as u8;
    }
    Ok(image)
}
